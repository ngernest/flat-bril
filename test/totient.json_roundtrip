{
  "functions": [
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "args": [
            "n"
          ],
          "dest": "tot",
          "funcs": [
            "totient"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "tot"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "n",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "n"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "p",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "label": "for.set.cond"
        },
        {
          "args": [
            "p",
            "p"
          ],
          "dest": "pp",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "pp",
            "n"
          ],
          "dest": "cond",
          "funcs": [],
          "labels": [],
          "op": "le",
          "type": "bool"
        },
        {
          "args": [
            "cond"
          ],
          "funcs": [],
          "labels": [
            "for.set.body",
            "for.set.end"
          ],
          "op": "br"
        },
        {
          "label": "for.set.body"
        },
        {
          "args": [
            "n",
            "p"
          ],
          "dest": "npmod",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "npmod",
            "zero"
          ],
          "dest": "if_cond",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "if_cond"
          ],
          "funcs": [],
          "labels": [
            "if_lbl",
            "else_lbl"
          ],
          "op": "br"
        },
        {
          "label": "if_lbl"
        },
        {
          "label": "while.set.cond"
        },
        {
          "args": [
            "n",
            "p"
          ],
          "dest": "npmod",
          "funcs": [
            "mod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "npmod",
            "zero"
          ],
          "dest": "while_cond",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "while_cond"
          ],
          "funcs": [],
          "labels": [
            "while.body",
            "while.end"
          ],
          "op": "br"
        },
        {
          "label": "while.body"
        },
        {
          "args": [
            "n",
            "p"
          ],
          "dest": "npdiv",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "npdiv"
          ],
          "dest": "n",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "while.set.cond"
          ],
          "op": "jmp"
        },
        {
          "label": "while.end"
        },
        {
          "args": [
            "result",
            "p"
          ],
          "dest": "resdiv",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "result",
            "resdiv"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "label": "else_lbl"
        },
        {
          "args": [
            "p",
            "one"
          ],
          "dest": "p",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.set.cond"
          ],
          "op": "jmp"
        },
        {
          "label": "for.set.end"
        },
        {
          "args": [
            "n",
            "one"
          ],
          "dest": "final_if_cond",
          "funcs": [],
          "labels": [],
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "final_if_cond"
          ],
          "funcs": [],
          "labels": [
            "final_if_label",
            "final_else_label"
          ],
          "op": "br"
        },
        {
          "label": "final_if_label"
        },
        {
          "args": [
            "result",
            "n"
          ],
          "dest": "resdiv",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "result",
            "resdiv"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "label": "final_else_label"
        },
        {
          "args": [
            "result"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "totient",
      "type": "int"
    },
    {
      "args": [
        {
          "name": "a",
          "type": "int"
        },
        {
          "name": "b",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "a",
            "b"
          ],
          "dest": "ad",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "b",
            "ad"
          ],
          "dest": "mad",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "a",
            "mad"
          ],
          "dest": "ans",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "ans"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "mod",
      "type": "int"
    }
  ]
}
