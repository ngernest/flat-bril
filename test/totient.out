@main
	op: print		args: ["n"]	
	op: call 		dest: tot		type: int		args: ["n"]		funcs: ["totient"]
	op: print		args: ["tot"]	
@totient
	op: id   		dest: result		type: int		args: ["n"]	
	op: const		dest: p 		type: int		value: 2	
	op: const		dest: one		type: int		value: 1	
	op: const		dest: zero		type: int		value: 0	
.for.set.cond
	op: mul  		dest: pp		type: int		args: ["p", "p"]	
	op: le   		dest: cond		type: bool		args: ["pp", "n"]	
	op: br   		args: ["cond"]		labels: ["for.set.body", "for.set.end"]
.for.set.body
	op: call 		dest: npmod		type: int		args: ["n", "p"]		funcs: ["mod"]
	op: eq   		dest: if_cond		type: bool		args: ["npmod", "zero"]	
	op: br   		args: ["if_cond"]		labels: ["if_lbl", "else_lbl"]
.if_lbl
.while.set.cond
	op: call 		dest: npmod		type: int		args: ["n", "p"]		funcs: ["mod"]
	op: eq   		dest: while_cond		type: bool		args: ["npmod", "zero"]	
	op: br   		args: ["while_cond"]		labels: ["while.body", "while.end"]
.while.body
	op: div  		dest: npdiv		type: int		args: ["n", "p"]	
	op: id   		dest: n 		type: int		args: ["npdiv"]	
	op: jmp  		labels: ["while.set.cond"]
.while.end
	op: div  		dest: resdiv		type: int		args: ["result", "p"]	
	op: sub  		dest: result		type: int		args: ["result", "resdiv"]	
.else_lbl
	op: add  		dest: p 		type: int		args: ["p", "one"]	
	op: jmp  		labels: ["for.set.cond"]
.for.set.end
	op: gt   		dest: final_if_cond		type: bool		args: ["n", "one"]	
	op: br   		args: ["final_if_cond"]		labels: ["final_if_label", "final_else_label"]
.final_if_label
	op: div  		dest: resdiv		type: int		args: ["result", "n"]	
	op: sub  		dest: result		type: int		args: ["result", "resdiv"]	
.final_else_label
	op: ret  		args: ["result"]	
@mod
	op: div  		dest: ad		type: int		args: ["a", "b"]	
	op: mul  		dest: mad		type: int		args: ["b", "ad"]	
	op: sub  		dest: ans		type: int		args: ["a", "mad"]	
	op: ret  		args: ["ans"]	
